File: python/waveq/glass.py
import os
import numpy as np

_JULIA_PROJECT = os.path.join(os.path.dirname(__file__), "..", "..", "src", "Waveq")
_JULIA_PROJECT = os.path.abspath(_JULIA_PROJECT)
os.environ.setdefault("JULIA_PROJECT", _JULIA_PROJECT)

from juliacall import Main as jl


class Glass:
    def __init__(self, jl_glass):
        self._jl = jl_glass

    @property
    def name(self):
        return str(self._jl.name)

    @property
    def wv_min(self):
        return float(self._jl.wv_min)

    @property
    def wv_max(self):
        return float(self._jl.wv_max)

    def refractive_index(self, wv):
        return float(jl.Waveq.refractive_index(self._jl, float(wv)))

    def refractive_index_array(self, wv_array):
        return np.array([self.refractive_index(w) for w in wv_array])


N_BK7 = Glass(jl.Waveq.N_BK7)
F2 = Glass(jl.Waveq.F2)
SF5 = Glass(jl.Waveq.SF5)

wv_ref_F = float(jl.Waveq.wv_ref_F)
wv_ref_e = float(jl.Waveq.wv_ref_e)
wv_ref_d = float(jl.Waveq.wv_ref_d)
wv_ref_C = float(jl.Waveq.wv_ref_C)

File: src/glass.jl
module glass

export Glass, refractive_index,
       wv_ref_F, wv_ref_e, wv_ref_d, wv_ref_C,
       N_BK7, F2, SF5

const wv_ref_F = 0.48613  # [um], H blue line
const wv_ref_e = 0.54607  # [um], Hg green line
const wv_ref_d = 0.58756  # [um], He yellow line
const wv_ref_C = 0.65627  # [um], H red line

struct Glass
    name::String
    wv_min::Float64
    wv_max::Float64
    B::Vector{Float64}
    C::Vector{Float64}
end

function refractive_index(g::Glass, wv::Real)::Float64
    return sqrt(1.0 + sum(g.B .* wv^2 ./ (wv^2 .- g.C)))
end

const N_BK7 = Glass(
    "N-BK7",
    0.365,
    2.5,
    [1.03961212, 0.231792344, 1.01046945],
    [0.00600069867, 0.0200179144, 103.560653],
)

const F2 = Glass(
    "F2",
    0.365,
    2.5,
    [1.34533359, 0.209073176, 0.937357162],
    [0.00997743871, 0.0470450767, 111.886764],
)

const SF5 = Glass(
    "SF5",
    0.365,
    2.5,
    [1.52481889, 0.187085527, 1.42729015],
    [0.011254756, 0.0588995392, 129.141675],
)

end # module
